### PR TITLE
feat(dashmate): reindex reworked

### PR DESCRIPTION
**Pull request by pshenmic**
## Issue being fixed or feature implemented

Dashmate reindex behaviour refactor.

Described in https://github.com/dashpay/platform/issues/788

## What was done?

* Removed complex logic that were starting separate container for reindex
* Removed containerId from the config
* Added `--detach` flag allowing to skip interactions during reindex command

## How Has This Been Tested?

* Locally on testnet node
* Locally on group nodes
* On devnet

## Breaking Changes

No

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by shumkov** [comment]
More information https://github.com/dashpay/platform/issues/788

File: packages/dashmate/src/commands/group/core/reindex.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ class GroupReindexCommand extends GroupBaseCommand {
     args,
     {
       verbose: isVerbose,
+      detach: isDetached,
     },
     isSystemConfig,
     reindexNodeTask,
```
Comment (shumkov) [inline packages/dashmate/src/commands/group/core/reindex.js:26]: We should ask only once and run reindex in parallel I guess (lister can do it)

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/src/commands/group/core/reindex.js [hunk 2]
```diff
@@ -56,7 +93,7 @@ class GroupReindexCommand extends GroupBaseCommand {
 
     try {
       await tasks.run({
-        isVerbose,
+        isDetached: true,
```
Comment (shumkov) [inline packages/dashmate/src/commands/group/core/reindex.js:96]: We could show the progress for all nodes, itsn't it?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/src/templates/renderServiceTemplatesFactory.js [hunk 3]
```diff
@@ -18,25 +17,20 @@ function renderServiceTemplatesFactory() {
...
 
-    const templatesPath = path.join(__dirname, '..', '..', 'templates');
-
     const templatePaths = glob
-      .sync(`${templatesPath}/**/*.dot`)
+      .sync(`${TEMPLATES_DIR}/**/*.dot`)
       // Do not render platform templates if it's not configured
       .filter((templatePath) => (
         !templatePath.includes('templates/platform') || config.get('platform.enable')
       ));
```
Comment (shumkov) [inline packages/dashmate/src/templates/renderServiceTemplatesFactory.js:25]: Do we still need this?

File: packages/dashmate/src/templates/writeServiceConfigsFactory.js [hunk 4]
```diff
@@ -20,7 +20,9 @@ function writeServiceConfigsFactory() {
     // Drop all files from configs directory
     const configDir = path.join(HOME_DIR_PATH, configName);
 
-    fs.rmSync(configDir, { recursive: true, force: true });
+    if (!fs.existsSync(configDir)) {
```
Comment (shumkov) [inline packages/dashmate/src/templates/writeServiceConfigsFactory.js:23]: Won't we create a directory bellow?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/src/templates/writeServiceConfigsFactory.js (hunk 4 above)
Comment (pshenmic) [reply packages/dashmate/src/templates/writeServiceConfigsFactory.js:23]: Yes. The logic was to purge the directory before render. I disabled removal and used check on folder existence instead

File: packages/dashmate/src/templates/renderServiceTemplatesFactory.js (hunk 3 above)
Comment (pshenmic) [reply packages/dashmate/src/templates/renderServiceTemplatesFactory.js:25]: Yes, why not?

File: packages/dashmate/src/templates/writeServiceConfigsFactory.js (hunk 4 above)
Comment (shumkov) [reply packages/dashmate/src/templates/writeServiceConfigsFactory.js:23]: I meant line 30-32:

```js
      // Recreate it
      fs.mkdirSync(absoluteFileDir, { recursive: true });
```

File: packages/dashmate/src/templates/renderServiceTemplatesFactory.js (hunk 3 above)
Comment (shumkov) [reply packages/dashmate/src/templates/renderServiceTemplatesFactory.js:25]: I mean currently all configs have `platform` section. If we remove delete then when we enable it we create files and then we disable it and we keep old version of templates. I think if we do not remove previous templates we should render all of them every time. What are your thoughts?

File: packages/dashmate/src/listr/tasks/reindexNodeTaskFactory.js [hunk 5]
```diff
@@ -38,125 +64,99 @@ function reindexNodeTaskFactory(
...
         },
       },
       {
-        title: 'Set reindex mode',
-        enabled: () => config.get('core.reindex.enable'),
+        title: 'Start Core in reindex mode',
         task: async () => {
-          config.set('core.reindex.enable', 1);
+          const templatePath = 'core/dash.conf.dot';
+          const [configPath] = templatePath.split('.dot');
```
Comment (shumkov) [inline packages/dashmate/src/listr/tasks/reindexNodeTaskFactory.js:96]: Looks confusing. Let's define configPath first and then use it as a part of templatePath?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/src/listr/tasks/reindexNodeTaskFactory.js [hunk 6]
```diff
@@ -38,125 +64,99 @@ function reindexNodeTaskFactory(
...
-            switch (State.ExitCode) {
-              default:
-              // eslint-disable-next-line no-console
-                console.warn(`Reindex container exited with status ${State.ExitCode}, check docker logs of container ${containerId}`);
-              // eslint-disable-next-line no-fallthrough
-              case 0:
-                await docker.getContainer(containerId).start();
-            }
+          if (isNodeRunning) {
+            return restartNodeTask(config);
```
Comment (shumkov) [inline packages/dashmate/src/listr/tasks/reindexNodeTaskFactory.js:116]: Why do we restart the whole node but not just the core service? Please add more comments to this function to explain logic and reasoning.

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/src/commands/group/core/reindex.js [hunk 7]
```diff
@@ -25,25 +27,60 @@ class GroupReindexCommand extends GroupBaseCommand {
...
+        task: async (ctx, task) => {
+          // Check if any of group nodes started
+          const isRunning = await configGroup
+            .reduce(async (acc, config) => (await acc || dockerCompose
+              .isNodeRunning(generateEnvs(configFile, config))), false);
+
+          let header;
+
+          if (isRunning) {
+            header = 'Node group running. The group nodes will be unavailable until reindex is complete.\n';
```
Comment (shumkov) [inline packages/dashmate/src/commands/group/core/reindex.js:47]:
```suggestion
            header = 'Node group is running. The group nodes will be unavailable until reindex is complete.\n';
```

File: packages/dashmate/src/listr/tasks/reindexNodeTaskFactory.js [hunk 8]
```diff
@@ -19,17 +22,40 @@ const generateEnvs = require('../../util/generateEnvs');
...
   configFile,
   getConnectionHost,
 ) {
+  /**
+   * Gets dashcore docker container from the node
+   * @param config
+   * @return {Promise<null|Container>}
+   */
+  async function getCoreContainer(config) {
+    const { docker } = dockerCompose;
```
Comment (shumkov) [inline packages/dashmate/src/listr/tasks/reindexNodeTaskFactory.js:43]: You should add docker as a dependency in `reindexNodeTaskFactory`

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/README.md [hunk 9]
```diff
@@ -283,19 +283,23 @@ rm -rf ~/.dashmate/
...
 
-The process displays interactive progress and can be interrupted at any time, but you cannot start your configuration until the resync is fully complete.
+The process displays interactive progress and may be interrupted at any time. After reindex is finished dashcore and other services will become online without any interactions from the user.
 
 The `core reindex` command works for regular and local configurations.
 
 ```
 USAGE
-  $ dashmate core reindex [-v] [--config <value>]
+  $ dashmate core reindex [-v] [--config <value>] [-d]
```
Comment (shumkov) [inline packages/dashmate/README.md:294]: Please update group and signle reindex command signatures

**Review by shumkov — CHANGES_REQUESTED** [review]

**Review by shumkov — APPROVED** [review]

**Review by shumkov — APPROVED** [review]

**Pull request merged**